ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -455,24 +444,194 @@ module Vendor = functor (V : VENDOR) -> struct
+  let find_one_by_name name t =
+    match find_by_name name t with
+    | x :: _ -> x
+    | [] -> failwith (Printf.sprintf "find_one_by_name(%s): not found" name)
+  let get_attr name = function
+    | E (_, attr, _) -> List.assoc name attr
+    | D _ -> failwith (Printf.sprintf "get_attr(%s): not an element" name)
+  let get_data = function 
+    | E (_, _, [D d]) -> d
+    | _ -> failwith "get_data: no data node"
+
+  let find_supported_vgpu_types device_id pgpus =
+    (*
+      Input example:
+        <pgpu>
+          <devId vendorId="0x10de" deviceId="0x1BB4" subsystemVendorId="0x10de" subsystemId="0x0"></devId>
+          <supportedVgpu vgpuId="72">
+            <maxVgpus>16</maxVgpus>
+            <digest type="signature">(...)</digest>
+          </supportedVgpu>
+          <supportedVgpu vgpuId="73">
+            (...)
+      Output: 
+      - Find the pgpus with deviceId = device_id (there's likely just one).
+      - List the vgpuIds of all supportedVgpus of these pgpus with their
+        maxVgpus value and the pgpu's subsystemId.
+    *)
+    List.filter_map (fun pgpu ->
+      let devid = find_one_by_name "devId" pgpu in
+      if int_of_string (get_attr "deviceId" devid) = device_id then
+        let psubdev_id =
+          let id = int_of_string (get_attr "subsystemId" devid) in
+          if id = 0 then None else Some id
+        in
+        let vgpus = find_by_name "supportedVgpu" pgpu in
+        Some (List.map (fun vgpu ->
+          let id = get_attr "vgpuId" vgpu in
+          let max = Int64.of_string (get_data (find_one_by_name "maxVgpus" vgpu)) in
+          id, (max, psubdev_id)
+        ) vgpus)
+      else
+        None
+    ) pgpus |> List.concat
+
+  let extract_conf whitelist device_id vgpu_types vgpu_ids =
+    (*
+      Input example:
+        <vgpuType id="11" name="GRID M60-0B" class="NVS">
+          <devId vendorId="0x10de" deviceId="0x13F2" subsystemVendorId="0x10de" subsystemId="0x1176"></devId>
+          <framebuffer>0x1A000000</framebuffer>
+          <numHeads>2</numHeads>
+          <display width="2560" height="1600"></display>
+          (...)
+        </vgpuType>
+      Output:
+      - Find the vgpuType elements that match the vgpu type ids found above,
+        and construct vgpu_conf records.
+    *)
+    List.filter_map (fun vgpu_type ->
+      let id = get_attr "id" vgpu_type in
+      if List.mem_assoc id vgpu_ids then
+        let max_instance, psubdev_id = List.assoc id vgpu_ids in
+        let framebufferlength = Int64.of_string (get_data (find_one_by_name "framebuffer" vgpu_type)) in
+        let num_heads = Int64.of_string (get_data (find_one_by_name "numHeads" vgpu_type)) in
+        let max_x, max_y =
+          let display = find_one_by_name "display" vgpu_type in
+          Int64.of_string (get_attr "width" display),
+          Int64.of_string (get_attr "height" display)
+        in
+        let devid = find_one_by_name "devId" vgpu_type in
+        let identifier = Identifier.({
```

Comment:
I believe this can be written as `Identifier.{..}`

Reply:
That's true.

Reply:
I've updated this after rebasing the PR.